lib/HdtDocument.cc

```diff
@@ -368,6 +369,111 @@ NAN_METHOD(HdtDocument::SearchTerms) {
     info[4]->IsObject() ? info[4].As<Object>() : info.This()));
 }
 
+/******** HdtDocument#_fetchDistinctTerms ********/
+
+class FetchDistinctTermsWorker : public Nan::AsyncWorker {
+  HdtDocument* document;
+  // JavaScript function arguments
+  string filterVal;
+  hdt::TripleComponentRole filterPos;
+  uint32_t limit;
+  hdt::TripleComponentRole position;
+  Persistent<Object> self;
+  // Callback return values
+  vector<string> distinctTerms;
+public:
+  FetchDistinctTermsWorker(HdtDocument* document, char* filterVal, uint32_t filterPos,
+                    uint32_t limit, uint32_t posId, Nan::Callback* callback,
+                    Local<Object> self)
+    : Nan::AsyncWorker(callback),
+      document(document), filterVal(filterVal), filterPos((TripleComponentRole)  filterPos),
+      limit(limit), position((TripleComponentRole) posId) {
+        SaveToPersistent("self", self);
+      };
+
+  void Execute() {
+
+    const char *subject = "";
+    const char *predicate = "";
+    const char *object = "";
+
+    // Initialize filter search term.
+    switch (filterPos) {
+      case SUBJECT:
+        subject = filterVal.c_str();
+        break;
+      case PREDICATE:
+        predicate = filterVal.c_str();
+        break;
+      case OBJECT:
+        object = filterVal.c_str();
+        break;
+    };
+
+    hdt::IteratorUCharString *it = NULL;
+    try {
+
+      Dictionary* dict = document->GetHDT()->getDictionary();
+      it = dict->getPredicates();
+
+      // Iterate over all predicates.
+      while (it->hasNext()) {
+        unsigned char *s = it->next();
+        predicate = reinterpret_cast<char*>(s);
+
+        // Check occurrence.
+        hdt::IteratorTripleString *rit = document->GetHDT()->search(subject, predicate, object);
+        if(rit->hasNext()) {
+          distinctTerms.push_back(predicate);
+        }
+
+        // Exit if reached limit.
+        if(limit && distinctTerms.size() == limit){
+          delete rit;
+          it->freeStr(s);
+          break;
+        }
+
+        delete rit;
+        it->freeStr(s);
+      }
```

Comment:
Repeated code above; can probably be mitigated by mitting the limit check in the while condition.